File: src/app/src/lib/monitoring/roboflow.ts
/**
 * Roboflow Workflow video client.
 *
 * Uses the Roboflow REST API (`POST /{workspace}/workflows/{workflow}`) to run workflow
 * object detection on uploaded video segments. The SDK's built-in WebRTC
 * transport (`webrtc.useVideoFile`) requires browser-level RTCPeerConnection
 * globals that are unavailable in Cloudflare Workers, so we bypass the SDK
 * entirely and call the serverless endpoint directly.
 */

import type { PluginWorkflowConfig } from "./plugins";

/** A normalized object detection result from the workflow. */
export interface WorkflowDetection {
  /** Detected class label (lowercase). */
  label: string;
  /** Confidence score (0-1). */
  confidence: number;
  /** Bounding box in absolute pixel coordinates (top-left + bottom-right). */
  box?: {
    xmin: number;
    ymin: number;
    xmax: number;
    ymax: number;
  };
  /** Timestamp in seconds within the video segment (if available). */
  atSec?: number;
  /** Frame index within the video (if available). */
  frameIndex?: number;
  /** Tracking ID for multi-frame tracking (if available). */
  trackId?: string;
  /** Roboflow class ID (if available). */
  classId?: number;
  /** Raw Roboflow prediction geometry (center-based coordinates and detection ID). */
  prediction?: {
    x: number;
    y: number;
    width: number;
    height: number;
    detectionId?: string;
  };
  /** Source image dimensions that the prediction coordinates are relative to. */
  image?: {
    width: number;
    height: number;
  };
}

/** Raw Roboflow workflow prediction shape. */
interface RawDetection {
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  confidence?: number;
  class?: string;
  class_id?: number;
  detection_id?: string;
  time?: number;
  timestamp?: number;
  frame_index?: number;
  frame_id?: number;
  tracker_id?: string;
  track_id?: string;
  [key: string]: unknown;
}

/** Runtime settings for the Roboflow workflow (shared across all plugins). */
export interface WorkflowRuntimeConfig {
  processingTimeoutSec: number;
  requestedPlan: string;
  requestedRegion: string;
}

/** Get shared runtime settings. */
export function getRuntimeConfig(): WorkflowRuntimeConfig {
  return {
    processingTimeoutSec: 3600,
    requestedPlan: "webrtc-gpu-medium",
    requestedRegion: "us",
  };
}

/** Options for running video object detection. */
export interface RunVideoDetectionOptions {
  /** The video segment bytes (MP4). */
  segmentBytes: Uint8Array;
  /** The workflow identity from the plugin catalog. */
  pluginWorkflow: PluginWorkflowConfig;
  /** The facility ID to get the container stub. */
  facilityId: string;
  /** Minimum confidence threshold for detections. */
  minConfidence?: number;
  /** If set, only these class labels are kept after detection. */
  classFilter?: string[];
  /** Frame sampling interval (default 30 = about 1 FPS for 30 FPS video). */
  frameInterval?: number;
  /** The Durable Object namespace for getting the container stub. */
  serverNamespace: DurableObjectNamespace;
  /** Roboflow API key from the Worker environment, forwarded to the Python container. */
  roboflowApiKey?: string;
  /** Roboflow API base URL from the Worker environment, forwarded to the Python container. */
  roboflowApiBase?: string;
}

/** Video metadata from the Python backend for playback alignment. */
export interface DetectionVideoMeta {
  fps: number;
  frameCount: number;
  decodedFrameCount?: number;
  frameInterval: number;
  attemptedFrameCount?: number;
  sampledFrameCount?: number;
  failedFrameCount?: number;
  skippedFrameCount?: number;
  circuitOpen?: boolean;
  processingDurationMs?: number;
  requestTimeoutSec?: number;
  queueWaitTimeoutSec?: number;
  videoBudgetSec?: number;
  maxConcurrency?: number;
}

/** A sampled frame's detection output. */
export interface DetectionFrame {
  /** Frame index within the video. */
  frameIndex: number;
  /** Timestamp in seconds. */
  atSec: number;
  /** Base64-encoded JPEG of the raw sampled frame. */
  beforeImage: string;
  /** Normalized detections for this frame. */
  detections: WorkflowDetection[];
  /** Source image dimensions that the detections are relative to. */
  image: { width: number; height: number };
}

/** Result of video object detection including metadata. */
export interface VideoDetectionResult {
  detections: WorkflowDetection[];
  detectionOutputs: DetectionFrame[];
  video: DetectionVideoMeta | null;
}

/**
 * Run a specific Roboflow workflow on a video segment via the Python backend.
 *
 * The Python backend (running in a Cloudflare Container) extracts frames using
 * OpenCV and calls the Roboflow REST API for each frame. This avoids the WebRTC
 * requirement that's unavailable in Cloudflare Workers.
 *
 * @returns Detections plus video metadata for playback alignment.
 */
export async function runVideoObjectDetection(options: RunVideoDetectionOptions): Promise<VideoDetectionResult> {
  const {
    segmentBytes,
    pluginWorkflow,
    facilityId,
    minConfidence = 0.4,
    classFilter,
    frameInterval = 30,
    serverNamespace,
    roboflowApiKey,
    roboflowApiBase,
  } = options;

  const containerStub = serverNamespace.getByName(facilityId) as ReturnType<DurableObjectNamespace["getByName"]> & {
    containerFetch(input: RequestInfo | URL, init?: RequestInit): Promise<Response>;
  };

  // Build the URL with query parameters
  const params = new URLSearchParams({
    workspace_name: pluginWorkflow.workspaceName,
    workflow_id: pluginWorkflow.workflowId,
    input_name: pluginWorkflow.inputName,
    frame_interval: String(frameInterval),
    min_confidence: String(minConfidence),
  });
  for (const outputName of pluginWorkflow.dataOutputNames ?? []) {
    params.append("data_output_names", outputName);
  }
  for (const label of classFilter ?? []) {
    params.append("class_filter", label);
  }

  const url = `http://localhost:3001/process-video?${params.toString()}`;
  const body = new ArrayBuffer(segmentBytes.byteLength);
  new Uint8Array(body).set(segmentBytes);

  // Frame retries happen inside the container. Replaying the complete video
  // here would repeat successful frame inference and amplify provider load.
  let response: Response;
  try {
    response = await containerStub.containerFetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "video/mp4",
        ...(roboflowApiKey ? { "X-Roboflow-Api-Key": roboflowApiKey } : {}),
        ...(roboflowApiBase ? { "X-Roboflow-Api-Base": roboflowApiBase } : {}),
      },
      body,
    });
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`Video detection request failed for workflow ${pluginWorkflow.workflowId}: ${message}`, {
      cause: error,
    });
  }

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Python backend error (${response.status}): ${errorText}`);
  }

  const result: unknown = await response.json();
  if (isRecord(result) && typeof result.error === "string") {
    throw new Error(`Python backend error: ${result.error}`);
  }
  let detections = parseWorkflowResponse(result, pluginWorkflow);

  detections = detections.filter((d) => d.confidence >= minConfidence);

  if (classFilter && classFilter.length > 0) {
    const allowedLabels = new Set(classFilter.map((label) => label.toLowerCase()));
    detections = detections.filter((d) => allowedLabels.has(d.label));
  }

  return {
    detections,
    detectionOutputs: extractDetectionOutputs(result),
    video: extractVideoMeta(result),
  };
}

// ── REST response parsing ──────────────────────────────────────────────────

/** Extract video metadata from Python backend response. */
function extractVideoMeta(result: unknown): DetectionVideoMeta | null {
  if (!result || typeof result !== "object") return null;
  const obj = result as Record<string, unknown>;
  const v = obj.video;
  if (!v || typeof v !== "object") return null;
  const video = v as Record<string, unknown>;
  if (
    typeof video.fps !== "number" ||
    typeof video.frameCount !== "number" ||
    typeof video.frameInterval !== "number"
  ) {
    return null;
  }
  return {
    fps: video.fps,
    frameCount: video.frameCount,
    decodedFrameCount: typeof video.decodedFrameCount === "number" ? video.decodedFrameCount : undefined,
    frameInterval: video.frameInterval,
    attemptedFrameCount: typeof video.attemptedFrameCount === "number" ? video.attemptedFrameCount : undefined,
    sampledFrameCount: typeof video.sampledFrameCount === "number" ? video.sampledFrameCount : undefined,
    failedFrameCount: typeof video.failedFrameCount === "number" ? video.failedFrameCount : undefined,
    skippedFrameCount: typeof video.skippedFrameCount === "number" ? video.skippedFrameCount : undefined,
    circuitOpen: typeof video.circuitOpen === "boolean" ? video.circuitOpen : undefined,
    processingDurationMs: typeof video.processingDurationMs === "number" ? video.processingDurationMs : undefined,
    requestTimeoutSec: typeof video.requestTimeoutSec === "number" ? video.requestTimeoutSec : undefined,
    queueWaitTimeoutSec: typeof video.queueWaitTimeoutSec === "number" ? video.queueWaitTimeoutSec : undefined,
    videoBudgetSec: typeof video.videoBudgetSec === "number" ? video.videoBudgetSec : undefined,
    maxConcurrency: typeof video.maxConcurrency === "number" ? video.maxConcurrency : undefined,
  };
}

/** Extract detection outputs from Python backend response. */
function extractDetectionOutputs(result: unknown): DetectionFrame[] {
  if (!result || typeof result !== "object") return [];
  const obj = result as Record<string, unknown>;
  const raw = obj.detectionOutputs;
  if (!Array.isArray(raw)) return [];

  return raw
    .filter(isRecord)
    .map((frame) => {
      if (typeof frame.frameIndex !== "number" || typeof frame.atSec !== "number") return null;
      if (typeof frame.beforeImage !== "string") return null;
      if (!Array.isArray(frame.detections)) return null;

      const image =
        isRecord(frame.image) && typeof frame.image.width === "number" && typeof frame.image.height === "number"
          ? { width: frame.image.width, height: frame.image.height }
          : { width: 0, height: 0 };

      const detections = frame.detections
        .filter(isRecord)
        .map((d) => passthroughDetection(d))
        .filter((d): d is WorkflowDetection => d !== null);

      return {
        frameIndex: frame.frameIndex,
        atSec: frame.atSec,
        beforeImage: frame.beforeImage,
        detections,
        image,
      };
    })
    .filter((f): f is DetectionFrame => f !== null);
}

/** Normalise the various shapes the Roboflow REST API may return. */
function parseWorkflowResponse(result: unknown, pluginWorkflow: PluginWorkflowConfig): WorkflowDetection[] {
  if (!result || typeof result !== "object") return [];

  const obj = result as Record<string, unknown>;

  // Python backend returns: { detections: [...], count: N }
  // The Python backend already returns the full normalized format including
  // raw/image fields, so pass them through directly.
  if (Array.isArray(obj.detections)) {
    return obj.detections
      .filter(isRecord)
      .map((d) => passthroughDetection(d))
      .filter((d): d is WorkflowDetection => d !== null);
  }

  // Roboflow REST API shapes (fallback):
  //   { outputs: [ { detections: [...], image: … } ] }
  //   { output:  { detections: [...], image: … }   }
  //   { detections: [...] }                          (single-image shorthand)
  const outputs: Record<string, unknown>[] = Array.isArray(obj.outputs)
    ? (obj.outputs as Record<string, unknown>[])
    : obj.output && typeof obj.output === "object"
      ? [obj.output as Record<string, unknown>]
      : Array.isArray(obj.detections)
        ? [obj as Record<string, unknown>]
        : [obj];

  const dataOutputNames = new Set(pluginWorkflow.dataOutputNames ?? [pluginWorkflow.inputName, "detections", "count"]);
  dataOutputNames.add("detections");
  dataOutputNames.add("predictions");

  const detections: WorkflowDetection[] = [];

  for (const output of outputs) {
    for (const [name, value] of Object.entries(output)) {
      if (name === "count") continue;
      if (!dataOutputNames.has(name)) continue;

      const rawDetections = extractDetections(value);
      // Capture image metadata from the detections output (same structure as Python parser)
      let imageMeta: WorkflowDetection["image"] | undefined;
      if (isRecord(value)) {
        const img = value.image as Record<string, unknown> | undefined;
        if (isRecord(img) && typeof img.width === "number" && typeof img.height === "number") {
          imageMeta = { width: img.width, height: img.height };
        }
      }

      for (const raw of rawDetections) {
        const detection = toDetection(raw, {}, imageMeta);
        if (detection) detections.push(detection);
      }
    }
  }

  return detections;
}

/**
 * Pass-through for detections already normalized by the Python backend.
 * Extracts the common fields and preserves raw/image metadata.
 */
function passthroughDetection(d: Record<string, unknown>): WorkflowDetection | null {
  if (typeof d.confidence !== "number") return null;
  return {
    label: String(d.label ?? "unknown"),
    confidence: d.confidence,
    box: d.box as WorkflowDetection["box"] | undefined,
    atSec: typeof d.atSec === "number" ? d.atSec : undefined,
    frameIndex: typeof d.frameIndex === "number" ? d.frameIndex : undefined,
    trackId: typeof d.trackId === "string" ? d.trackId : undefined,
    classId: typeof d.classId === "number" ? d.classId : undefined,
    prediction: (d.prediction ?? d.detection) as WorkflowDetection["prediction"] | undefined,
    image: d.image as WorkflowDetection["image"] | undefined,
  };
}

function extractDetections(value: unknown): RawDetection[] {
  if (Array.isArray(value)) return value.filter(isRecord) as RawDetection[];
  if (!isRecord(value)) return [];

  if (Array.isArray(value.value)) return value.value.filter(isRecord) as RawDetection[];
  if (isRecord(value.value)) return extractDetections(value.value);

  if (Array.isArray(value.detections)) return value.detections.filter(isRecord) as RawDetection[];
  if (isRecord(value.detections)) return extractDetections(value.detections);

  if (Array.isArray(value.predictions)) return value.predictions.filter(isRecord) as RawDetection[];
  if (isRecord(value.predictions)) return extractDetections(value.predictions);

  return [];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object";
}

/** Normalize a raw Roboflow workflow prediction into our shared detection shape. */
function toDetection(
  p: RawDetection,
  fallback: { frameIndex?: number; atSec?: number },
  imageMeta?: WorkflowDetection["image"],
): WorkflowDetection | null {
  if (typeof p.confidence !== "number") return null;

  const label = String(p.class ?? "unknown").toLowerCase();
  const confidence = p.confidence;

  let box: WorkflowDetection["box"] | undefined;
  let rawLocal: WorkflowDetection["prediction"] | undefined;
  if (
    typeof p.x === "number" &&
    typeof p.y === "number" &&
    typeof p.width === "number" &&
    typeof p.height === "number"
  ) {
    box = {
      xmin: p.x - p.width / 2,
      ymin: p.y - p.height / 2,
      xmax: p.x + p.width / 2,
      ymax: p.y + p.height / 2,
    };
    rawLocal = {
      x: p.x,
      y: p.y,
      width: p.width,
      height: p.height,
      detectionId: typeof p.detection_id === "string" ? p.detection_id : undefined,
    };
  }

  return {
    label,
    confidence,
    box,
    atSec: typeof p.time === "number" ? p.time : typeof p.timestamp === "number" ? p.timestamp : fallback.atSec,
    frameIndex:
      typeof p.frame_index === "number"
        ? p.frame_index
        : typeof p.frame_id === "number"
          ? p.frame_id
          : fallback.frameIndex,
    trackId: typeof p.tracker_id === "string" ? p.tracker_id : typeof p.track_id === "string" ? p.track_id : undefined,
    classId: typeof p.class_id === "number" ? p.class_id : undefined,
    prediction: rawLocal,
    image: imageMeta,
  };
}

File: src/server/roboflow.py
"""Roboflow workflow video processing via REST API.

Extracts frames from a video segment using OpenCV and sends each frame
to the Roboflow serverless workflow endpoint. Aggregates detections
into a normalized detection format.

Each sampled frame produces a detection output containing:
- beforeImage: the raw sampled frame (JPEG base64)
- detections: the raw detections from the workflow
- image: the inference image dimensions
"""

from __future__ import annotations

import base64
import logging
import math
import os
import tempfile
from typing import Any

import cv2
import httpx
import numpy as np

log = logging.getLogger("facilix.roboflow")

ROBOFLOW_API_BASE = os.getenv("ROBOFLOW_API_BASE", "https://serverless.roboflow.com")
ROBOFLOW_API_KEY = os.getenv("ROBOFLOW_API_KEY", "")


async def process_video_workflow(
    video_bytes: bytes,
    workspace_name: str,
    workflow_id: str,
    input_name: str = "image",
    data_output_names: list[str] | None = None,
    class_filter: list[str] | None = None,
    frame_interval: int = 30,
    min_confidence: float = 0.4,
    roboflow_api_key: str | None = None,
    roboflow_api_base: str | None = None,
) -> dict[str, Any]:
    """Process a video segment through a Roboflow workflow.

    Extracts frames at regular intervals and sends each to the workflow endpoint.
    Returns detections, detection outputs with frame images, and video metadata.
    """
    api_key = roboflow_api_key or ROBOFLOW_API_KEY
    api_base = (roboflow_api_base or ROBOFLOW_API_BASE).rstrip("/")

    if not api_key:
        raise ValueError("ROBOFLOW_API_KEY is not configured")
    if frame_interval < 1:
        raise ValueError("frame_interval must be at least 1")
    if not 0 <= min_confidence <= 1:
        raise ValueError("min_confidence must be between 0 and 1")

    output_names = set(data_output_names or [input_name, "detections", "predictions", "count"])
    output_names.update({"detections", "predictions"})
    allowed_labels = {label.lower() for label in class_filter or []}

    # Write video to a temporary file for OpenCV
    with tempfile.NamedTemporaryFile(suffix=".mp4", delete=False) as tmp:
        tmp.write(video_bytes)
        tmp_path = tmp.name

    try:
        cap = cv2.VideoCapture(tmp_path)
        if not cap.isOpened():
            raise ValueError("Failed to open video file")

        reported_fps = cap.get(cv2.CAP_PROP_FPS)
        fps = reported_fps if math.isfinite(reported_fps) and reported_fps > 0 else 30.0
        reported_frame_count = max(0, int(cap.get(cv2.CAP_PROP_FRAME_COUNT)))
        log.info(
            "processing video: reported %d frames @ %.1f fps, extracting every %d frames",
            reported_frame_count,
            fps,
            frame_interval,
        )

        detections: list[dict[str, Any]] = []
        detection_outputs: list[dict[str, Any]] = []
        frame_idx = 0
        attempted_frames = 0
        failed_frames = 0
        last_frame: np.ndarray | None = None
        last_frame_index = -1

        async with httpx.AsyncClient(timeout=60.0) as client:
            try:
                while True:
                    ret, frame = cap.read()
                    if not ret:
                        break

                    last_frame = frame
                    last_frame_index = frame_idx

                    if frame_idx % frame_interval == 0:
                        attempted_frames += 1
                        output = await _process_frame(
                            client=client,
                            frame=frame,
                            workspace_name=workspace_name,
                            workflow_id=workflow_id,
                            input_name=input_name,
                            output_names=output_names,
                            frame_index=frame_idx,
                            at_sec=frame_idx / fps,
                            min_confidence=min_confidence,
                            allowed_labels=allowed_labels,
                            api_key=api_key,
                            api_base=api_base,
                        )
                        if output:
                            detections.extend(output["detections"])
                            detection_outputs.append(output)
                        else:
                            failed_frames += 1

                    frame_idx += 1

                # OpenCV's reported frame count can be missing or inaccurate.
                # Use the last frame we actually decoded instead of seeking by metadata.
                if last_frame is not None and last_frame_index % frame_interval != 0:
                    attempted_frames += 1
                    output = await _process_frame(
                        client=client,
                        frame=last_frame,
                        workspace_name=workspace_name,
                        workflow_id=workflow_id,
                        input_name=input_name,
                        output_names=output_names,
                        frame_index=last_frame_index,
                        at_sec=last_frame_index / fps,
                        min_confidence=min_confidence,
                        allowed_labels=allowed_labels,
                        api_key=api_key,
                        api_base=api_base,
                    )
                    if output:
                        detections.extend(output["detections"])
                        detection_outputs.append(output)
                    else:
                        failed_frames += 1
            finally:
                cap.release()

        if frame_idx == 0:
            raise ValueError("Video contains no decodable frames")
        if attempted_frames > 0 and failed_frames == attempted_frames:
            raise RuntimeError(f"Roboflow inference failed for all {attempted_frames} sampled frames")

        log.info(
            "processed %d frames, %d detections, %d detection outputs (%d inference failures)",
            frame_idx,
            len(detections),
            len(detection_outputs),
            failed_frames,
        )

        return {
            "detections": detections,
            "count": len(detections),
            "detectionOutputs": detection_outputs,
            "video": {
                "fps": round(fps, 2),
                "frameCount": frame_idx,
                "frameInterval": frame_interval,
                "sampledFrameCount": len(detection_outputs),
                "failedFrameCount": failed_frames,
            },
        }

    finally:
        os.unlink(tmp_path)


async def _process_frame(
    client: httpx.AsyncClient,
    frame: np.ndarray,
    workspace_name: str,
    workflow_id: str,
    input_name: str,
    output_names: set[str],
    frame_index: int,
    at_sec: float,
    min_confidence: float,
    allowed_labels: set[str],
    api_key: str,
    api_base: str,
) -> dict[str, Any] | None:
    """Send a single frame to the Roboflow workflow and return a detection output dict.

    Returns None if the request fails. The dict contains:
    - frameIndex, atSec: frame timing
    - beforeImage: raw sampled frame as JPEG base64
    - detections: normalized detections for this frame
    - image: inference image dimensions
    """
    # Encode frame as JPEG (before image)
    _, before_buf = cv2.imencode(".jpg", frame)
    before_b64 = base64.b64encode(before_buf).decode("utf-8")

    url = f"{api_base}/{workspace_name}/workflows/{workflow_id}"

    try:
        resp = await client.post(
            url,
            json={
                "api_key": api_key,
                "inputs": {
                    input_name: {"type": "base64", "value": before_b64},
                },
            },
        )

        if resp.status_code != 200:
            log.warning(
                "roboflow frame %d failed: HTTP %d — %s",
                frame_index,
                resp.status_code,
                resp.text[:200],
            )
            return None

        result = resp.json()
        frame_data = _parse_frame_response(
            result,
            output_names,
            frame_index,
            at_sec,
            min_confidence,
            allowed_labels,
            {"width": int(frame.shape[1]), "height": int(frame.shape[0])},
        )

        return {
            "frameIndex": frame_index,
            "atSec": round(at_sec, 3),
            "beforeImage": before_b64,
            "detections": frame_data["detections"],
            "image": frame_data["image"],
        }

    except Exception as exc:
        log.exception("roboflow frame %d error: %s", frame_index, exc)
        return None


def _parse_frame_response(
    result: Any,
    output_names: set[str],
    frame_index: int,
    at_sec: float,
    min_confidence: float,
    allowed_labels: set[str],
    default_image_meta: dict[str, int],
) -> dict[str, Any]:
    """Parse the Roboflow REST API response into detections + image metadata.

    Returns an output even if no valid detections are found so sampled frames
    can still be persisted for debugging/review.
    """
    # Roboflow Workflows commonly returns a top-level list. Direct model
    # responses and some workflow versions return an object instead.
    if isinstance(result, list):
        outputs = result
    elif isinstance(result, dict):
        outputs = result.get("outputs") or result.get("output") or [result]
    else:
        outputs = []
    if not isinstance(outputs, list):
        outputs = [outputs]

    raw_detections: list[dict[str, Any]] = []
    image_meta: dict[str, int] | None = None

    for output in outputs:
        if not isinstance(output, dict):
            continue

        for name, value in output.items():
            if name == "count" or name not in output_names:
                continue

            # Capture prediction image metadata (dimensions the detections are relative to)
            if image_meta is None and isinstance(value, dict) and isinstance(value.get("image"), dict):
                img = value["image"]
                if isinstance(img.get("width"), (int, float)) and isinstance(img.get("height"), (int, float)):
                    image_meta = {"width": int(img["width"]), "height": int(img["height"])}

            raw_detections.extend(_extract_detections(value))

    if image_meta is None:
        image_meta = default_image_meta

    detections: list[dict[str, Any]] = []
    for pred in raw_detections:
        detection = _to_detection(pred, frame_index, at_sec, image_meta)
        if (
            detection
            and detection["confidence"] >= min_confidence
            and (not allowed_labels or detection["label"] in allowed_labels)
        ):
            detections.append(detection)

    return {"detections": detections, "image": image_meta}


def _extract_detections(value: Any) -> list[dict[str, Any]]:
    """Recursively extract detections from various response formats."""
    if isinstance(value, list):
        return [p for p in value if isinstance(p, dict)]

    if not isinstance(value, dict):
        return []

    # Try common nested keys
    for key in ("value", "detections", "predictions"):
        nested = value.get(key)
        if isinstance(nested, list):
            return [p for p in nested if isinstance(p, dict)]
        if isinstance(nested, dict):
            return _extract_detections(nested)

    return []


def _to_detection(
    pred: dict[str, Any],
    frame_index: int,
    at_sec: float,
    image_meta: dict[str, int] | None = None,
) -> dict[str, Any] | None:
    """Normalize a Roboflow prediction into our detection format.

    Returns both normalized bounding boxes and raw prediction geometry
    so the frontend can render boxes and use source-image dimensions.
    """
    confidence = pred.get("confidence")
    if not isinstance(confidence, (int, float)):
        return None

    label = str(pred.get("class", "unknown")).lower()

    # Convert center-based bbox to corner-based
    x = pred.get("x")
    y = pred.get("y")
    width = pred.get("width")
    height = pred.get("height")
    has_geometry = all(isinstance(v, (int, float)) for v in [x, y, width, height])

    box = None
    if has_geometry:
        box = {
            "xmin": x - width / 2,
            "ymin": y - height / 2,
            "xmax": x + width / 2,
            "ymax": y + height / 2,
        }

    detection = {
        "label": label,
        "confidence": float(confidence),
        "box": box,
        "atSec": at_sec,
        "frameIndex": frame_index,
        "trackId": pred.get("tracker_id") or pred.get("track_id"),
        "classId": pred.get("class_id"),
        # Raw prediction geometry so frontend can render with source-image coordinates
        "prediction": {
            "x": x,
            "y": y,
            "width": width,
            "height": height,
            "detectionId": pred.get("detection_id"),
        }
        if has_geometry
        else None,
    }

    if image_meta:
        detection["image"] = image_meta

    return detection
